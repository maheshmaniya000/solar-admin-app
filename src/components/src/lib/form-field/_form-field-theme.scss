@use '~@angular/material' as mat;

@mixin theme($theme-or-color-config) {
  $color-config: mat.get-color-config($theme-or-color-config);
  $primary: map-get($color-config, primary);
  $warn: map-get($color-config, warn);
  $foreground: map-get($color-config, foreground);
  $is-dark-theme: map-get($color-config, is-dark);
  $underline-color-base: mat.get-color-from-palette($foreground, divider, if($is-dark-theme, 1, 0.87));
  $underline-focused-color: mat.get-color-from-palette($primary, text);

  mat-form-field.mat-form-field-appearance-outline {
    .mat-form-field-subscript-wrapper .mat-hint {
      color: mat.get-color-from-palette($foreground, hint-text);
    }

    &,
    &.mat-form-field-disabled {
      .mat-form-field-outline {
        color: mat.get-color-from-palette($foreground, disabled);
      }
    }

    &,
    &.mat-focused {
      .mat-form-field-outline-thick {
        color: mat.get-color-from-palette($foreground, icon);
      }
    }

    &.mat-form-field-invalid {
      &[doNotAppendFormError] {
        .mat-form-field-outline-thick,
        &.mat-focused .mat-form-field-outline-thick {
          color: $underline-focused-color;
        }
      }

      ::placeholder {
        color: mat.get-color-from-palette($warn);
      }
    }
  }
}
