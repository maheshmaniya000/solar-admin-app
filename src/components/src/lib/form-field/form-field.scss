@use '../icon/icon' as icon;

$horizontal-spacing: 8px;

mat-form-field {
  width: 100%;

  &.mat-form-field {
    display: block;
  }

  &.mat-form-field-appearance-outline {
    ::placeholder {
      transition: none;
      font-style: italic;
    }

    .mat-form-field-wrapper {
      margin: 0;
      padding-bottom: 0;
    }

    .mat-form-field-outline {
      top: 0;
    }

    .mat-form-field-outline-start {
      border-radius: 4px 0 0 4px;
    }

    .mat-form-field-outline-end {
      border-radius: 0 4px 4px 0;
    }

    .mat-form-field-outline-thick {
      .mat-form-field-outline-start,
      .mat-form-field-outline-end,
      .mat-form-field-outline-gap {
        border-width: 1px;
      }
    }

    &[doNotAppendFormError].mat-form-field-invalid {
      &:not(.mat-focused) .mat-form-field-outline {
        opacity: 1;

        &.mat-form-field-outline-thick {
          opacity: 0;
        }
      }
    }

    .mat-form-field-flex {
      align-items: center;
      margin-top: 0;
      padding: 0 $horizontal-spacing;
    }

    .mat-form-field-infix {
      padding: 5px 0;
      border-top: none;
    }

    .mat-form-field-subscript-wrapper {
      position: static;
      top: 0;
      margin-top: 4px;
      padding: 0;
      font-size: 100%;
    }

    &.ng-untouched,
    &.ng-valid,
    &.ng-pending {
      &.no-hint .mat-form-field-subscript-wrapper {
        display: none;
      }
    }

    .mat-form-field-prefix,
    .mat-form-field-suffix {
      position: static;

      .mat-icon {
        @include icon.size(default);
      }
    }

    &:not(.mat-form-field-appearance-legacy) {
      .mat-form-field-prefix,
      .mat-form-field-suffix {
        .mat-icon-button {
          display: block;
          width: auto;
          height: auto;
        }
      }
    }

    .mat-form-field-prefix {
      margin-right: $horizontal-spacing;
    }

    .mat-form-field-suffix {
      margin-left: $horizontal-spacing;
    }

    .mat-form-field-label-wrapper {
      display: none;
    }
  }
}
