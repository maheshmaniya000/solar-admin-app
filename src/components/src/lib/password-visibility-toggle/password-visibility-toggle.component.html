<button type="button" mat-icon-button (click)="onToggle()">
  <mat-icon>{{ icon }}</mat-icon>
</button>
