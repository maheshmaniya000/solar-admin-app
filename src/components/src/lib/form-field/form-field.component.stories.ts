import { APP_BASE_HREF, CommonModule } from '@angular/common';
import { HttpClientModule } from '@angular/common/http';
import { FormsModule } from '@angular/forms';
import { MatButtonModule } from '@angular/material/button';
import { MatIconModule } from '@angular/material/icon';
import { MatInputModule } from '@angular/material/input';
import { BrowserAnimationsModule } from '@angular/platform-browser/animations';
import { moduleMetadata, Story } from '@storybook/angular';

import { TranslocoRootModule } from 'ng-shared/core/transloco-root.module';

import { SgDefaultsModule } from '../defaults/defaults.module';
import { FormErrorModule } from '../form-error/form-error.module';
import { FormFieldModule } from './form-field.module';

export default {
  title: 'Solargis/Components/Form Field',
  decorators: [
    moduleMetadata({
      imports: [
        BrowserAnimationsModule,
        MatInputModule,
        SgDefaultsModule,
        FormFieldModule,
        FormsModule,
        CommonModule,
        MatIconModule,
        MatButtonModule,
        HttpClientModule,
        TranslocoRootModule,
        FormErrorModule
      ],
      providers: [{ provide: APP_BASE_HREF, useValue: '.' }]
    })
  ]
};

export const bottomSpacing: Story = props => ({
  props,
  template: `
    <div class="container">
      <label for="input1">Hint</label>
      <div>
        <mat-form-field>
          <input id="input1" matInput placeholder="placeholder">
          <mat-hint *ngIf="hintVisible">Some hint</mat-hint>
        </mat-form-field>
        <div>Some block to demonstrate form field height without hint.</div>
      </div>

      <label for="input2">Error</label>
      <div>
        <mat-form-field>
          <input id="input2" matInput placeholder="placeholder" ngModel required>
        </mat-form-field>
        <div>
          Some block to demonstrate form field height with dynamic error. Since
          this field is required, you can see its error when its value is empty,
          and you interact with the input at least once (click in and out).
        </div>
      </div>

      <label for="input3">Error + hint</label>
      <div>
        <mat-form-field>
          <input id="input3" matInput placeholder="placeholder" ngModel required>
          <mat-hint *ngIf="hintVisible">Some hint</mat-hint>
        </mat-form-field>
        <div>
          Some block to demonstrate form field height with dynamic error & hint. Since
          this field is required, you can see its error when its value is empty,
          and you interact with the input at least once (click in and out).
        </div>
      </div>
    </div>
  `,
  styles: [
    `
    .container {
      display: grid;
      grid-template-columns: repeat(2, 1fr);
      grid-gap: 10px;
      width: 800px;
    }
  `
  ]
});

bottomSpacing.argTypes = {
  hintVisible: {
    name: 'Hint visible',
    control: { type: 'boolean' },
    defaultValue: false
  }
};

export const disabled: Story = props => ({
  props,
  template: `
    <div class="container">
      <p>
        'Required' error message will be rendered only if this input is touched
         and enabled. To demonstrate that disabled inputs do not display error, first
         interact with it (click in and out) and then set the disabled control to
         'True'.
      </p>
      <mat-form-field>
        <input matInput ngModel [disabled]="disabled" required placeholder="placeholder"/>
      </mat-form-field>
    </div>
  `,
  styles: [
    `
    .container {
      width: 500px;
    }
  `
  ]
});

disabled.argTypes = {
  disabled: {
    name: 'Disabled',
    control: { type: 'boolean' },
    defaultValue: false
  }
};

export const prefixAndSuffix: Story = () => ({
  template: `
    <div class="container">
      <label for="input1">Text</label>
      <mat-form-field>
        <input id="input1" matInput placeholder="placeholder"/>
        <span matPrefix>+421</span>
        <span matSuffix>kWh</span>
      </mat-form-field>
      <label for="input2">Icon / Icon button</label>
      <mat-form-field>
        <input id="input2" matInput placeholder="placeholder"/>
        <mat-icon matPrefix>search</mat-icon>
        <button mat-icon-button color="primary" matSuffix>
          <mat-icon>visibility_off</mat-icon>
        </button>
      </mat-form-field>
    </div>
  `,
  styles: [
    `
    .container {
      display: grid;
      grid-template-columns: repeat(2, 1fr);
      grid-gap: 10px;
      width: 500px;
    }
  `
  ]
});
