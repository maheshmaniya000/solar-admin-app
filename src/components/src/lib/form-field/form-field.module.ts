import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { MatFormFieldModule } from '@angular/material/form-field';
import { TranslocoModule } from '@ngneat/transloco';

import { NoHintDirective } from './no-hint/no-hint.directive';

@NgModule({
  declarations: [NoHintDirective],
  imports: [CommonModule, MatFormFieldModule, TranslocoModule],
  exports: [MatFormFieldModule, NoHintDirective]
})
export class FormFieldModule {}
