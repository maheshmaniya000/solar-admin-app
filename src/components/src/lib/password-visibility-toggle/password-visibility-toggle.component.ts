import {
  ChangeDetectionStrategy,
  Component,
  Host,
  OnInit,
  SkipSelf
} from '@angular/core';
import { MatFormField } from '@angular/material/form-field';
import { isNil } from 'lodash-es';

@Component({
  selector: 'sg-password-visibility-toggle',
  templateUrl: './password-visibility-toggle.component.html',
  changeDetection: ChangeDetectionStrategy.OnPush
})
export class PasswordVisibilityToggleComponent implements OnInit {
  icon: 'visibility' | 'visibility_off';

  private input: HTMLInputElement;

  constructor(
    @Host() @SkipSelf() private readonly matFormField: MatFormField
  ) {}

  ngOnInit(): void {
    // eslint-disable-next-line no-underscore-dangle
    const matFormFieldElement: HTMLElement = this.matFormField._elementRef.nativeElement;
    this.input = matFormFieldElement.querySelector('input');

    if (isNil(this.input)) {
      throw new Error('Password visibility toggle must be used with input');
    }

    this.updateIcon();
  }

  onToggle(): void {
    this.input.type = this.isInputTypePassword() ? 'text' : 'password';
    this.updateIcon();
  }

  private updateIcon(): void {
    this.icon = this.isInputTypePassword() ? 'visibility_off' : 'visibility';
  }

  private isInputTypePassword(): boolean {
    return this.input.type === 'password';
  }
}
