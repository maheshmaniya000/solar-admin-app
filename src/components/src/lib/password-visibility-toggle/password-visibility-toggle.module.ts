import { NgModule } from '@angular/core';
import { MatButtonModule } from '@angular/material/button';
import { MatIconModule } from '@angular/material/icon';

import { PasswordVisibilityToggleComponent } from './password-visibility-toggle.component';

@NgModule({
  declarations: [PasswordVisibilityToggleComponent],
  imports: [MatIconModule, MatButtonModule],
  exports: [PasswordVisibilityToggleComponent]
})
export class PasswordVisibilityToggleModule {}
