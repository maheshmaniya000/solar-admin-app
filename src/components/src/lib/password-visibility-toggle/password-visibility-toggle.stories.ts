import { MatInputModule } from '@angular/material/input';
import { BrowserAnimationsModule } from '@angular/platform-browser/animations';
import { moduleMetadata, Story } from '@storybook/angular';

import { SgDefaultsModule } from '../defaults/defaults.module';
import { FormFieldModule } from '../form-field/form-field.module';
import { PasswordVisibilityToggleModule } from './password-visibility-toggle.module';

export default {
  title: 'Solargis/Components/Password Visibility Toggle',
  decorators: [
    moduleMetadata({
      imports: [
        BrowserAnimationsModule,
        SgDefaultsModule,
        FormFieldModule,
        MatInputModule,
        PasswordVisibilityToggleModule
      ]
    })
  ]
};

export const basic: Story = props => ({
  props,
  template: `
    <mat-form-field>
      <input matInput type="password"/>
      <sg-password-visibility-toggle matSuffix>
      </sg-password-visibility-toggle>
    </mat-form-field>
  `
});
