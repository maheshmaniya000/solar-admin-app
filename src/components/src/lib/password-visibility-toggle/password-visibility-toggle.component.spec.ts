import { Component, ViewChild } from '@angular/core';
import { MatFormField, MatSuffix } from '@angular/material/form-field';
import { MatInput } from '@angular/material/input';
import { createComponentFactory, Spectator } from '@ngneat/spectator/jest';
import { MockDirective } from 'ng-mocks';

import { PasswordVisibilityToggleComponent } from './password-visibility-toggle.component';

import spyOn = jest.spyOn;
import Mock = jest.Mock;

describe('PasswordVisibilityToggleComponent', () => {
  describe('unit', () => {
    let component: PasswordVisibilityToggleComponent;
    let querySelectorMock: Mock;
    let matFormFieldStub: Partial<MatFormField>;

    const setupTest = (matFormField: Partial<MatFormField>): void => {
      component = new PasswordVisibilityToggleComponent(
        matFormField as MatFormField
      );
      component.ngOnInit();
    };

    beforeEach(() => {
      querySelectorMock = jest.fn();
      matFormFieldStub = {
        // eslint-disable-next-line @typescript-eslint/naming-convention
        _elementRef: {
          nativeElement: {
            querySelector: querySelectorMock
          }
        }
      };
    });

    describe('ngOnInit', () => {
      it('should throw an error when MatFormField does not contain input', () => {
        querySelectorMock.mockReturnValue(undefined);
        expect(() => setupTest(matFormFieldStub)).toThrow(/input/);
      });

      [
        { type: 'password', expectedIcon: 'visibility_off' },
        { type: 'text', expectedIcon: 'visibility' }
      ].forEach(({ type, expectedIcon }) =>
        it(`should set icon to '${expectedIcon}' when input type is '${type}'`, () => {
          querySelectorMock.mockReturnValue({ type });
          setupTest(matFormFieldStub);
          expect(component.icon).toBe(expectedIcon);
        })
      );
    });

    describe('onToggle', () => {
      let inputStub: Partial<HTMLInputElement>;

      beforeEach(() => {
        inputStub = { type: 'password' };
        querySelectorMock.mockReturnValue(inputStub);
        setupTest(matFormFieldStub);
      });

      it('should toggle input type', () => {
        component.onToggle();
        expect(inputStub.type).toBe('text');
        component.onToggle();
        expect(inputStub.type).toBe('password');
      });

      it('should toggle icon', () => {
        component.onToggle();
        expect(component.icon).toBe('visibility');
        component.onToggle();
        expect(component.icon).toBe('visibility_off');
      });
    });
  });

  describe('component', () => {
    @Component({
      selector: 'sg-test',
      template: `
        <mat-form-field>
          <input matInput type="password" />
          <sg-password-visibility-toggle matSuffix>
          </sg-password-visibility-toggle>
        </mat-form-field>
      `
    })
    class TestComponent {
      @ViewChild(PasswordVisibilityToggleComponent)
      passwordVisibilityComponent: PasswordVisibilityToggleComponent;
    }

    let spectator: Spectator<TestComponent>;

    const createComponent = createComponentFactory({
      component: TestComponent,
      declarations: [
        PasswordVisibilityToggleComponent,
        MatFormField,
        MatInput,
        MockDirective(MatSuffix)
      ],
      shallow: true
    });

    beforeEach(() => {
      spectator = createComponent();
    });

    it('should match snapshot and have bound all properties and methods', () => {
      const onToggleSpy = spyOn(
        spectator.component.passwordVisibilityComponent,
        'onToggle'
      );
      spectator.click('sg-password-visibility-toggle button');
      expect(onToggleSpy).toHaveBeenCalledTimes(1);

      expect(spectator.fixture).toMatchSnapshot();
    });
  });
});
