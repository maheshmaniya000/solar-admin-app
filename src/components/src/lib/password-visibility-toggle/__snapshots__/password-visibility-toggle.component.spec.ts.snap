// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PasswordVisibilityToggleComponent component should match snapshot and have bound all properties and methods 1`] = `
<sg-test
  __ngContext__={[Function LRootView]}
  passwordVisibilityComponent={[Function PasswordVisibilityToggleComponent]}
>
  <mat-form-field
    class="mat-form-field mat-primary mat-form-field-type-mat-input mat-form-field-appearance-legacy mat-form-field-can-float mat-form-field-hide-placeholder _mat-animation-noopable mat-form-field-should-float mat-focused"
  >
    <div
      class="mat-form-field-wrapper"
    >
      <div
        class="mat-form-field-flex"
      >
        <div
          class="mat-form-field-infix"
        >
          <input
            aria-invalid="false"
            aria-required="false"
            class="mat-input-element mat-form-field-autofill-control cdk-text-field-autofill-monitored"
            matinput=""
            type="text"
          />
          <span
            class="mat-form-field-label-wrapper"
          >
          </span>
        </div>
        <div
          class="mat-form-field-suffix"
        >
          <sg-password-visibility-toggle
            matsuffix=""
          >
            <button
              mat-icon-button=""
              type="button"
            >
              <mat-icon>
                visibility
              </mat-icon>
            </button>
          </sg-password-visibility-toggle>
        </div>
      </div>
      <div
        class="mat-form-field-underline"
      >
        <span
          class="mat-form-field-ripple"
        />
      </div>
      <div
        class="mat-form-field-subscript-wrapper"
      >
        <div
          class="mat-form-field-hint-wrapper ng-trigger ng-trigger-transitionMessages"
          style="opacity:1;transform:translateY(0%);0:opacity;1:transform;"
        >
          <div
            class="mat-form-field-hint-spacer"
          />
        </div>
      </div>
    </div>
  </mat-form-field>
</sg-test>
`;
